driver/driver.go

```diff
@@ -217,3 +246,123 @@ func newSafeMounter() *mount.SafeFormatAndMount {
+		state[machine.ID] = machine.Disks
+	}
+	return state, nil
+}
+
+func (d *Driver) runOVCStatemachine() {
+	var state map[int][]int
+	var ac attachConfig
+
+	defer d.done.Done()
+
+	newStateMachine := func() {
+		var err error
+		d.log.Info("Creating ovc state machine")
+		for {
+			if state, err = d.createStateInventory(); err == nil {
+				break
+			}
+			d.log.Info("Failed to create state machine. Retrying in 30 seconds")
+			time.Sleep(30 * time.Second)
+		}
+	}
+
+	indexOf := func(slice []int, value int) int {
+		for i, v := range slice {
+			if v == value {
+				return i
+			}
+		}
+		return -1
+	}
+
+	remove := func(slice []int, i int) []int {
+		slice[len(slice)-1], slice[i] = slice[i], slice[len(slice)-1]
+		return slice[:len(slice)-1]
+	}
+
+	attach := func() error {
+		for machineID, disks := range state {
+			if index := indexOf(disks, ac.diskID); index >= 0 {
+				if machineID == ac.machineID {
+					d.log.Infof("Nothing to do, this disk %d is already attached to machine %d", ac.diskID, machineID)
+					ac.result <- nil
+					return nil
+				}
+				// Disk is attached to the wrong machine: disconnect
+				if err := d.client.Disks.Detach(&ovc.DiskAttachConfig{
+					MachineID: machineID,
+					DiskID:    ac.diskID,
+				}); err != nil {
+					d.log.Errorf("Failed to detach disk %d from machine %d: %s", ac.diskID, machineID, err)
+					ac.result <- err
+					return err
+				}
+				d.log.Infof("Detached disk %d from machine %d", ac.diskID, machineID)
+				state[machineID] = remove(disks, index)
+				break
+			}
+		}
+		// Attach the disk to the correct machine
+		if err := d.client.Disks.Attach(&ovc.DiskAttachConfig{
+			MachineID: ac.machineID,
+			DiskID:    ac.diskID,
+		}); err != nil {
+			d.log.Errorf("Failed to attach disk %d to machine %d: %s", ac.diskID, ac.machineID, err)
+			ac.result <- err
+			return err
+		}
+		state[ac.machineID] = append(state[ac.machineID], ac.diskID)
+		ac.result <- nil
+		d.log.Errorf("Attached disk %d to machine %d", ac.diskID, ac.machineID)
+		return nil
+	}
+
+	detach := func() error {
+		for machineID, disks := range state {
+			if index := indexOf(disks, ac.diskID); index >= 0 {
+				if err := d.client.Disks.Detach(&ovc.DiskAttachConfig{
+					MachineID: machineID,
+					DiskID:    ac.diskID,
+				}); err != nil {
+					d.log.Errorf("Failed to detach disk %d from machine %d: %s", ac.diskID, machineID, err)
+					ac.result <- err
+					return err
+				}
+				d.log.Infof("Detached disk %d from machine %d", ac.diskID, machineID)
+				state[machineID] = remove(disks, index)
+				break
+			}
+		}
+		ac.result <- nil
+		return nil
+	}
+
+	newStateMachine()
+	for {
+		select {
+		case ac = <-d.attach:
+			if err := attach(); err != nil {
+				d.log.Info("Error while executing attach request. Recycling state machine")
+				newStateMachine()
+			}
+		case ac = <-d.detach:
+			if err := detach(); err != nil {
+				d.log.Info("Error while executing detach request. Recycling state machine")
```

Comment:
Log as warning or error, also log the error message itself

Reply:
Same here